Feature/14 adapt request event notification

spresser-hlrs opened the pull request:
Resolves issue #14 by adding a class that allows bidirectional conversion between SLURM- and Torque-style notification settings. This also adds constants SCHEDULER_SLURM and SCHEDULER_TORQUE which are exported and can be used to represent the schedulers instead of passing around strings.

Testing:
The tests for this are in tests/test_jobfile_generator.py. These verify:

- That the conversion between the two formats works
- That no conversion occurs if an option is unrecognized
- That no conversion occurs when no arguments are passed
- That no conversion occurs when the format already matches the scheduler in use

Tests must be run inside the docker container.

kmlTE commented on MODAK/src/jobfile_generator.py:
```diff
@@ -3,6 +3,75 @@
+            # convert, then return
+            return self._slurm_to_torque(notifications)
+        # Scheduler and options match; return the original notification options
+        # (Also, if notifications are invalid this leaves them intact/unmodified)
```
If notifications are invalid, the job script will also be invalid and won't run. In this case, I think we should not include notifications in the job script if it is invalid.

spresser-hlrs replied:
Would you like me to filter them if they're invalid? This code currently allows invalid notifications to pass so that it doesn't change how they're handled. If I change this to filter invalid notifications, it will change the behavior of MODAK as a whole.

kmlTE commented on MODAK/src/jobfile_generator.py:
```diff
@@ -3,6 +3,75 @@
+
+class ArgumentConverter():
+        # Define which options are vaild of each of the schedulers
+    VAILD_TORQUE=['', 'a','b','e'] # abort, begin, end
+    VALID_SLURM=["", "NONE", "BEGIN", "END", "FAIL", "REQUEUE", "ALL",
```
Empty string is invalid format and it won't run the job script

kmlTE commented on MODAK/src/jobfile_generator.py:
```diff
@@ -96,7 +166,8 @@ def __generate_torque_header(self):
             f.write(DIRECTIVE + ' -W ' + self.job_data['job_dependency'])
             f.write('\n')
         if "request_event_notification" in self.job_data:
-            f.write(DIRECTIVE + ' -m ' + self.job_data['request_event_notification'])
+            f.write(DIRECTIVE + ' -m ' + 
+                self.ac.convert_notifications(SCHEDULER_TORQUE, self.job_data['request_event_notification']))
```
I think we should check for validity first and then not include if invalid

kmlTE commented on MODAK/src/jobfile_generator.py:
```diff
@@ -187,7 +258,8 @@ def __generate_slurm_header(self):
             f.write(DIRECTIVE + ' --dependency=' + self.job_data['job_dependency'])
             f.write('\n')
         if "request_event_notification" in self.job_data:
-            f.write(DIRECTIVE + ' --mail-type=' + self.job_data['request_event_notification'])
+            f.write(DIRECTIVE + ' --mail-type=' + 
+                self.ac.convert_notifications(SCHEDULER_SLURM, self.job_data['request_event_notification']))
```
Same as for PBS

spresser-hlrs commented:
Okay, I have updated this code so that in the event of invalid notification settings, it will replace the invalid settings with reasonable defaults. I also removed the blank options as valid settings. Finally, I added handling for the Torque f, n, and p options.
